File: DiffUtilTest/app/src/main/java/com/example/diffutiltest/viewmodel/StudentViewModel.java
package com.example.diffutiltest.viewmodel;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.diffutiltest.entity.Student;

import java.util.List;

/**
 * create by hy on 2019/11/19 23:02
 */
public class StudentViewModel extends ViewModel {

    MutableLiveData<List<Student>> studentLiveData = new MutableLiveData<>();

    public MutableLiveData<List<Student>> getStudentLiveData() {
        return studentLiveData;
    }

    public void setStudents(List<Student> students) {
        Log.d("StudentViewModel", "onClick: size = " + students.size());
        studentLiveData.setValue(students);
    }
}

File: KotlinDemos/StudentList/app/src/main/java/com/example/studentlist/viewmodels/StudentViewModel.kt
package com.example.studentlist.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.studentlist.db.StudentRepository
import com.example.studentlist.entity.Student
import com.example.studentlist.utils.LogUtils
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.IllegalArgumentException

class StudentViewModel(private val repository: StudentRepository) : ViewModel() {
    val allStudent: LiveData<List<Student>> = repository.allStudent

    fun insert(student: Student) = viewModelScope/*GlobalScope*/.launch {
        LogUtils.d("insert currentThread=${Thread.currentThread().name}")
        repository.insert(student)
    }
}

class StudentViewModelFactory(private val repository: StudentRepository) :
    ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(StudentViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return StudentViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }

}